### DISCUSSION TITLE
Getting started questions

### DISCUSSION THREAD

**Discussion by oyvindberg** [oyvindberg/bleep#354] [Q&A]
Hi!

This is an open discussion for doubts people have when trying out bleep. I'll later process the information from here and merge it with the documentation on the next documentation pass.

Also feel free to start new topics if you suspect the answer is going to be long :)

**Comment by dwalend** [comment]
How do I build just the .jar file ? (Not an executable .jar file with all the dependencies - just the jar file to use as a library.)

Thanks,

David

**Reply by oyvindberg** [reply]
Publish-local?

**Reply by dwalend** [reply]
That wants some extra bits:

`
13:42 $ bleep publish-local Ev3LangScala Ev3LangScalaExample
📗 bootstrapped in 46 ms
Missing expected flag --groupId!
Missing expected flag --version!
`

I really just need the .jar file - to ssh onto a different machine. (Plus I need something more explicitly clear for my students.)

**Reply by oyvindberg** [reply]
Don't punish them with a jar file without a source jar! How about cloning a repo with the files you want to share? Or publish your code so they can depend on it like a normal jar?

**Reply by dwalend** [reply]
They need to scp the .jar file to a Lego EV3. The drive on the robot's end is running at a pretty slow speed and the kids have a short attention span, so the less they have to copy the better. The .jar file contains exactly their code; the libraries it uses are all already there. 

They need just the minimal .jar to save time copying so to avoid wasting their youth. The robot can't even start the scala compiler, and can only show 44 characters on the screen. Not a lot of reason to have source code there.

**Reply by oyvindberg** [reply]
Aha, you're not gonna share it. 

Well, the closest is `bleep dist` and `bleep publish-local` for now. I suppose we could add a `bleep package` or `bleep jar` command

**Reply by dwalend** [reply]
I could do beep jar. (You've already built all of it except the named command.) Where should I start in the code?

**Reply by dwalend** [reply]
(I'll also be doing an OS-independent ssh and scp if you want them - but those are part of a very specialized command.)

**Reply by oyvindberg** [reply]
make a new command in `bleep-core/src/scala/bleep/commands` and wire it up in `bleep-cli/src/scala/bleep/Main.scala`.

we could talk about ssh and scp if there is a need, I'm sure interesting things could be built in top of it

**Reply by dwalend** [reply]
Thanks! I'll see what I can do.

**Reply by dwalend** [reply]
I'm building .jar files - as a plug-in. I get the guiding idea of making what you need, and agree that the world doesn't need to make .jars as a library without sharing them. (I need it, followed by scp, so the plugin makes sense.)

Is there a reason why buildDist doesn't make an executable .jar file (with all the dependencies baked in)?

**Reply by dwalend** [reply]
I've got bleep building .jar files and moving them around using scp. 

Does bleep have a good way of handling passwords? 

Does bleep supply an API to get to the .yaml file contents - where I can define my own key/value pairs?

Thanks!

**Reply by oyvindberg** [reply]
> Is there a reason why buildDist doesn't make an executable .jar file (with all the dependencies baked in)?

This is commonly called a `fatjar` on the JVM. It's fairly arbitrary that bleep build a directory with jars and a script instead. The code I ported happened to do it that way. It's also a more simple process, because you don't need to merge files this way.

> Does bleep have a good way of handling passwords?
> 
> Does bleep supply an API to get to the .yaml file contents - where I can define my own key/value pairs?

this is exactly where bleep ends and bleep scripts start - write a bleep script, and it's infinitely customizable because it's just code.

**Reply by dwalend** [reply]
Is there some API - maybe in Started - that will let me access the config values in bleep.yaml ?

I'd like to keep my plugins' config bits there - to have just one config file.

**Reply by oyvindberg** [reply]
No, they are not read. See #272

**Reply by dwalend** [reply]
Thanks. I'm reading my own .yaml file, which isn't so bad. 

Next Question: Is there a way to pass jvm-style -D properties into bleep? (I want to pass an argument in to my script - to tell it which hostname to scp to.) (Interestingly, trying -Dkey=val does not error, but does not pass in the value. However -Ekey-val gets an "Unexpected option: -E" . Is this a bug?)

**Reply by dwalend** [reply]
More on passing -D system properties: It looks like they don't get passed forward in Run.scala 's .rawRun() . Should they be? (Is this even the right code path?)

`
  def rawRun(started: Started, bloop: BuildServer, main: String): Either[BleepException, Unit] =
    Compile(watch = false, Array(project)).runWithServer(started, bloop).map { case () =>
      cli(
        "run",
        started.pre.buildPaths.cwd,
        jvmRunCommand(started.bloopProject(project), started.resolvedJvm, project, Some(main), args).orThrow,
        logger = started.logger,
        out = cli.Out.Raw,
        in = cli.In.Attach,
        env = sys.env.toList
      )
      ()
    }
`

**Reply by oyvindberg** [reply]
Not entirely sure. it would be nice if you could try and make a small table of how it actually works now, and especially if there is something which works if done in the correct way.

`rawRun` is used when running `bleep run`. this means bleep starts the JVM itself, and is in complete control. the other option is to go through bloop, which has some implications with stdin/out, env, props and all these things. I haven't tested it exhaustively, but I added the `raw` option because something was broken with terminal handling when going through bloop.

running with bloop is used for scripts afair, at least until there is a reason to change it. 

i thought both and and java props would be forwarded, but if they are not it would be useful to figure out why. note that bleep will consume some input parameters unless you preceed them with `--`.

**Reply by dwalend** [reply]
I read a bit more about bloop's test target and found a combination that works. Thanks for pointing me to that!

The `--` has to go last, and it needs `-J-Dkey=value` .

`bleep scpJar Ev3LangScalaExample -- -J-DrobotName=firefly` for example.

**Comment by oyvindberg** [comment]
> Thanks. I'm reading my own .yaml file, which isn't so bad.
it's a little bit bad, because when bleep writes back to the file (as it's meant to do for instance when updating deps, or for a bunch of commands I haven't implemented yet) the changes will be lost.

**Comment by dwalend** [comment]
I have a project that has several dependencies on the same version of a fleet of libraries. What's a good way to avoid the cut/paste of the version number?

```
    dependencies:
      - org.http4s::http4s-ember-server:0.23.27
      - org.http4s::http4s-ember-client:0.23.27
      - org.http4s::http4s-circe:0.23.27
      - org.http4s::http4s-dsl:0.23.27

```

Thanks,

David

**Reply by oyvindberg** [reply]
You duplicate them. And then you update deps with `bleep build update-deps`, so you don't miss any.

Try it for a while, I really enjoy the clarity of doing it this way.

Note that you can share whole (groups of) dependencies with templates

**Reply by dwalend** [reply]
Thanks! ... Is there something I can go read about templates? (I've been messing with Smithy4s and I think there's enough there to rate its own template.)

**Comment by dwalend** [comment]
Any advice for handling that for things with multiple active versions (cough h4s cough) ? Bleep wants to switch my project to their the latest experimental stuff.

**Reply by oyvindberg** [reply]
try the newest version of bleep, it's meant to use the same algorithm as scala steward. i use http4s, and it reliably bumps me to newest 0.23.x
